content/md/posts/2023-07-08-testing.md

```diff
@@ -0,0 +1,150 @@
+{:title "Starting out with testing"
+ :layout :post
+ :tags  ["cljs" "testing" "re-frame"]
+ :toc true}
+
+## Overview
+
+This article covers why I want to have testing in my application, how to do the setup for the type of tests I'll be doing, and what future testing posts will be about.
+
+This post will discuss what is needed beyond the setup procedure described in "setting up the project part 1" and "setting up the project part two" create the tests. Please follow the steps in the previous posts before continuing on with this post.
+
+## Why do I want testing in my application?
+
+For my application, I want to know when it is working. Instead of having to wait for something bad to happen so I learn that something has gone wrong, it would be nice to check beforehand so I can then diagnose the problem.
+
+I started searching online for ways to accomplish this, and testing was the first thing that showed up. From what I read, testing is used to check a variety of things and acts as an alarm of sorts. The test fails, and from there you can investigate the problem.
+
+The three types of tests which I saw were unit, integration, and acceptance testing.
+
+### Unit tests
+
+Unit tests take a function and checks the result of the function given a specific input. It focuses mainly on testing that one component, and only that component, works well.
+
+For example, in my project I have to setup the database when starting everything up. 
+I have a function below that returns a map with the default values of the database
+
+```clojure
+(re-frame/reg-event-db
+ ::initialize-db
+ (fn-traced [_ _]
+            db/default-db))
+```
+
+
+I want to test that this function returns the correct default database. To do this I made a file in test/haggadah called events_test.cljs. In the namespace I added a few dependencies for the testing framework I use and to the file where I store my events. It should look like this.
+
+```clojure
+(ns haggadah.events-test
+  (:require [haggadah.subs :as subs]
+            [cljs.test :as t :include-macros true]
+            [re-frame.core :as rf]
+            [day8.re-frame.test :as rf-test]
+            [haggadah.events :as events]
+            [haggadah.core :as core]
+            [haggadah.routes :as routes]))
+```
+
+For testing that `::initialize-db` returns default-db, which is just a map of the form `{:name nil :uid nil :user :unloaded}`,
```

Comment:
NIT: Usually unit testing doesn't involve events, etc... perhaps you can extract the _init_ function and call it from the event handler, and show that function (that just returns a map) with a `deftest` and `testing`.

Reply:
I did not realize that. I'll change the test so it just uses functions.